### PR TITLE
[Feat] add PHPStan

### LINKED ISSUE #135: PHPStan
We should implement PHPStan and have GitHub Actions enforce it.

### PR DESCRIPTION
**Pull request by joelbutcher**
This PR resolves #135 and adds PHPStan to the project.

## Action Explanation
This PR adds two new CI actions. The first is simple, it's the static analysis pipeline to ensure PHPStan is satisfied prior to merging the PR.

The second job is more nuanced.

Because we're using a combination of matrix job variants, presumably some form of branch protection rules and multiple CI jobs we don't technically have an exposed way to know if all CI has passed. If **any** CI pipeline or job variant fails, GitHubs internal `check` job doesn't get marked as `failed`, it gets marked as `skipped` (why it does this, I have no clue!).

The new `check` job (which has to be called this to override the default job run by GitHub) always runs once any of its dependents (`test` and `static-analysis`) completes, regardless of whether that was a successful completion or failed completion. This is useful, as it will give @jbrooksuk a more reliable source of truth as to the merge state of a PR. I.e. if `check` fails, we must not merge this PR.

### CODE REVIEW

**Comment by jbrooksuk** [comment]
@joelbutcher thanks for doing this! I've just fixed a few of the issues in `main` itself, since they're simple and will lower the amount of base lining we need to do. Can you update accordingly?

**Comment by joelbutcher** [comment]
@jbrooksuk I'm now getting new errors:

```
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  Line   src/Http/Resources/ComponentGroup.php                                                                                                                                                     
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  31     Method Cachet\Http\Resources\ComponentGroup::toRelationships() should return array<string, (callable():                                                                                   
         TiMacDonald\JsonApi\JsonApiResource)|Illuminate\Http\Resources\PotentiallyMissing|TiMacDonald\JsonApi\JsonApiResourceCollection> but returns array{components: Closure(): (iterable<int,  
         mixed>&TiMacDonald\JsonApi\JsonApiResourceCollection)}.                                                                                                                                   
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 

 ------ ------------------------------------------------------------------------------ 
  Line   src/Http/Resources/Incident.php                                               
 ------ ------------------------------------------------------------------------------ 
  48     Call to static method make() on an unknown class Cachet\Http\Resources\User.  
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols           
 ------ ------------------------------------------------------------------------------ 
```

**Comment by jbrooksuk** [comment]
Okay, that should exist now.

File: .github/workflows/run-tests.yml [hunk 1]
```diff
@@ -47,3 +85,18 @@ jobs:
...
+      - static-analysis
+      - test
+
+    runs-on: Ubuntu-latest
+
+    steps:
+      - name: Check if all jobs are passing
+        uses: re-actors/alls-green@release/v1
+        with:
+          jobs: ${{ toJSON(needs) }}
```
Comment (jbrooksuk) [inline .github/workflows/run-tests.yml:102]: @joelbutcher what is this doing exactly?

**Review by jbrooksuk — COMMENTED** [review]
Looks good to me as a good starting point. Just curious about this GitHub action change.

File: .github/workflows/run-tests.yml (hunk 1 above)
Comment (joelbutcher) [reply .github/workflows/run-tests.yml:102]: @jbrooksuk Ah yes, sorry! I was supposed to add this into the PR description!

You should use this job as the "required check" that enables PR's in branch protection to determine if a PR is mergeable.

The way GitHub does this is weird when you have multiple CI pipelines, as well as using a matrix. GitHub doesn't mark the `check` job as `failed` it marks it as `skipped`.

The `if: always()` bit ensures this check is **always** run even if one of the `test` job matrix variants. or the `static-analysis` jobs fail.

Comment (jbrooksuk) [reply .github/workflows/run-tests.yml:102]: Love it!

**Comment by jbrooksuk** [comment]
I made some changes to the `run-tests` action, it's now running but failing — https://github.com/cachethq/core/actions/runs/12185288288

**Comment by joelbutcher** [comment]
@jbrooksuk Strange that the action didn't run as part of this PR. I'll take a look at this later today when I have some spare time

**Comment by joelbutcher** [comment]
@jbrooksuk Looks like a lot of the errors I thought I had fixed, but I hadn't pushed the commit up 🙈. Sending in a follow up PR now :)

**Comment by jbrooksuk** [comment]
Thanks @joelbutcher!

**Pull request merged**